### ISSUE TITLE
Crash Psalm

### ISSUE THREAD

**Issue by orklah** [vimeo/psalm#3337]
After solving a dependancy issue on our project, we get a crash on psalm:

```
Uncaught TypeError: Argument 2 passed to Psalm\Internal\Provider\NodeDataProvider::setType() must be an instance of Psalm\Type\Union, instance of stdClass given, called in C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\Fetch\VariableFetchAnalyzer.php on line 277 and defined in C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Provider\NodeDataProvider.php:41

Stack trace:

#0 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\Fetch\VariableFetchAnalyzer.php(277): Psalm\Internal\Provider\NodeDataProvider->setType(Object(PhpParser\Node\Expr\Variable), Object(stdClass))

#1 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer.php(103): Psalm\Internal\Analyzer\Statements\Expression\Fetch\VariableFetchAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\Variable), Object(Psalm\Context), false, NULL, false)

#2 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\CallAnalyzer.php(811): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\Variable), Object(Psalm\Context))

#3 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\CallAnalyzer.php(451): Psalm\Internal\Analyzer\Statements\Expression\CallAnalyzer::handleByRefFunctionArg(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), 'preg_match', 2, Object(PhpParser\Node\Arg), Object(Psalm\Context))

#4 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\Call\FunctionCallAnalyzer.php(294): Psalm\Internal\Analyzer\Statements\Expression\CallAnalyzer::checkFunctionArguments(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Array, Array, 'preg_match', Object(Psalm\Context))

#5 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer.php(442): Psalm\Internal\Analyzer\Statements\Expression\Call\FunctionCallAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\FuncCall), Object(Psalm\Context))

#6 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\LoopAnalyzer.php(604): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\FuncCall), Object(Psalm\Context))

#7 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\LoopAnalyzer.php(336): Psalm\Internal\Analyzer\Statements\Block\LoopAnalyzer::applyPreConditionToLoopContext(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\FuncCall), Array, Object(Psalm\Context), Object(Psalm\Context), false)

#8 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\WhileAnalyzer.php(60): Psalm\Internal\Analyzer\Statements\Block\LoopAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Array, Array, Array, Object(Psalm\Internal\Scope\LoopScope), Object(Psalm\Context))

#9 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(398): Psalm\Internal\Analyzer\Statements\Block\WhileAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\While_), Object(Psalm\Context))

#10 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\FunctionLikeAnalyzer.php(541): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context), NULL, true)

#11 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\ClassAnalyzer.php(1767): Psalm\Internal\Analyzer\FunctionLikeAnalyzer->analyze(Object(Psalm\Context), Object(Psalm\Internal\Provider\NodeDataProvider), NULL)

#12 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\ClassAnalyzer.php(727): Psalm\Internal\Analyzer\ClassAnalyzer->analyzeClassMethod(Object(PhpParser\Node\Stmt\ClassMethod), Object(Psalm\Storage\ClassLikeStorage), Object(Psalm\Internal\Analyzer\ClassAnalyzer), Object(Psalm\Context), NULL)

#13 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(845): Psalm\Internal\Analyzer\ClassAnalyzer->analyze(Object(Psalm\Context), NULL)

#14 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\IfAnalyzer.php(795): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context))

#15 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\IfAnalyzer.php(348): Psalm\Internal\Analyzer\Statements\Block\IfAnalyzer::analyzeIfBlock(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Internal\Scope\IfScope), Object(Psalm\Context), Object(Psalm\Context), Object(Psalm\Context), Array)

#16 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(382): Psalm\Internal\Analyzer\Statements\Block\IfAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Context))

#17 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\FileAnalyzer.php(190): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context), NULL, true)

#18 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\IncludeAnalyzer.php(158): Psalm\Internal\Analyzer\FileAnalyzer->analyze(Object(Psalm\Context), false, NULL)

#19 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer.php(743): Psalm\Internal\Analyzer\Statements\Expression\IncludeAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\Include_), Object(Psalm\Context), NULL)

#20 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(751): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\Include_), Object(Psalm\Context), false, NULL, true)

#21 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\IfAnalyzer.php(795): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context))

#22 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Block\IfAnalyzer.php(348): Psalm\Internal\Analyzer\Statements\Block\IfAnalyzer::analyzeIfBlock(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Internal\Scope\IfScope), Object(Psalm\Context), Object(Psalm\Context), Object(Psalm\Context), Array)

#23 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(382): Psalm\Internal\Analyzer\Statements\Block\IfAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Context))

#24 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\FileAnalyzer.php(190): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context), NULL, true)

#25 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\IncludeAnalyzer.php(158): Psalm\Internal\Analyzer\FileAnalyzer->analyze(Object(Psalm\Context), false, NULL)

#26 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer.php(743): Psalm\Internal\Analyzer\Statements\Expression\IncludeAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\Include_), Object(Psalm\Context), NULL)

#27 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(751): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\Include_), Object(Psalm\Context), false, NULL, true)

#28 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\FileAnalyzer.php(190): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context), NULL, true)

#29 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Codebase\Analyzer.php(357): Psalm\Internal\Analyzer\FileAnalyzer->analyze(NULL)

#30 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Codebase\Analyzer.php(589): Psalm\Internal\Codebase\Analyzer->Psalm\Internal\Codebase\{closure}(5948, 'C:\\Users\\{redacted}...')

#31 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Codebase\Analyzer.php(267): Psalm\Internal\Codebase\Analyzer->doAnalysis(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 1)

#32 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\ProjectAnalyzer.php(582): Psalm\Internal\Codebase\Analyzer->analyzeFiles(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 1, false, true)

#33 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\src\psalm.php(588): Psalm\Internal\Analyzer\ProjectAnalyzer->check('C:\\Users\\{redacted}...', false)

#34 C:\Users\{redacted}\PhpstormProjects\vendor\vimeo\psalm\psalm(2): require_once('C:\\Users\\{redacted}...')

#35 {main}

(Psalm dev-master@733da0e53935c50f997340340b41b5bc8ca83016 crashed due to an uncaught Throwable)
```

When launching with --debug-by-line, it crash on `unset($parse);` on the code below:

```php
/**
* parser le code HTML
*
* @param string    contenu a parser.
* @param    &array    tableau de retour des donnees
* @return    null
*/
function searchCode($content, &$tmp) {
    // separer les balises du texte
    $tmp = [];
    $reg = '/(<[^>]+>)|([^<]+)+/isU';

    // pour chaque element trouve :
    $str    = '';
    $offset = 0;
    while(preg_match($reg, $content, $parse, PREG_OFFSET_CAPTURE, $offset)) {
        // si une balise a ete detectee
        if($parse[1][0]) {
            // sauvegarde du texte precedent si il existe
            if($str !== '') {
                $tmp[] = ['txt', $str];
            }

            // sauvegarde de la balise
            $tmp[] = ['code', trim($parse[1][0])];

            // initialisation du texte suivant
            $str = '';
        } else {
            // ajout du texte a la fin de celui qui est deja detecte
            $str .= $parse[2][0];
        }
        // Update offset to the end of the match
        $offset = $parse[0][1] + strlen($parse[0][0]);
        unset($parse);
    }
    // si un texte est present a la fin, on l'enregistre
    if($str != '') {
        $tmp[] = ['txt', $str];
    }
    unset($str);
}
```

Pasting the method as-is on psalm.dev does not reproduce the issue. Note that this is a dependancy on our project so it is ignored in our xml file config

**Comment by muglug** [comment]
Fixed in 3554aa4

**Comment by orklah** [comment]
Thanks!
